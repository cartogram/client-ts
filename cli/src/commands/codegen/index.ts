import { generateFromContext } from '@xata.io/codegen';
import chalk from 'chalk';
import { mkdir, writeFile } from 'fs/promises';
import fetch from 'node-fetch';
import path, { dirname, extname, relative } from 'path';
import { BaseCommand, ProjectConfig } from '../../base.js';

const languages: Record<string, 'javascript' | 'typescript'> = {
  '.js': 'javascript',
  '.mjs': 'javascript',
  '.ts': 'typescript'
};

export default class Codegen extends BaseCommand {
  static description = 'Generate code form the current database schema';

  static examples = [];

  static flags = {};

  static args = [];

  async run(): Promise<void> {
    const output = this.projectConfig?.codegen?.output;

    if (!output) {
      return this.error(
        `Please, specify an output file in your project configuration file first with ${chalk.bold(
          'xata config set codegen.output <path>'
        )}`
      );
    }

    const ext = extname(output);
    const dir = dirname(output);
    const language = languages[ext];
    if (!language) {
      return this.error(
        `Cannot generate code for a file with extension ${ext}. Please use one of the following extensions: ${Object.keys(
          languages
        ).join(', ')}`
      );
    }

    const databaseURL = await this.getDatabaseURL();
    const result = await generateFromContext('typescript', { fetchImpl: fetch, databaseURL });
    const code = result.transpiled;
    const declarations = result.declarations;

    await mkdir(dir, { recursive: true });
    await writeFile(output, code);
    if (declarations && this.projectConfig?.codegen?.declarations) {
      await writeFile(path.join(dir, 'types.d.ts'), declarations);
    }

    this.log(`Your XataClient is generated at ./${relative(process.cwd(), output)}`);
  }

  static async runIfConfigured(projectConfig: ProjectConfig) {
    if (projectConfig?.codegen?.output) return Codegen.run([]);
  }
}
