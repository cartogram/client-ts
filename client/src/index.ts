export interface XataRecord {
  _id: string;
  _version: number;
  read(): Promise<this>;
  update(data: Selectable<this>): Promise<this>;
  delete(): Promise<void>;
}

export type Queries<T> = {
  [key in keyof T as T[key] extends Query<infer A, infer B> ? key : never]: T[key];
};

export type OmitQueries<T> = {
  [key in keyof T as T[key] extends Query<infer A, infer B> ? never : key]: T[key];
};

export type OmitLinks<T> = {
  [key in keyof T as T[key] extends XataRecord ? never : key]: T[key];
};

export type OmitMethods<T> = {
  // eslint-disable-next-line @typescript-eslint/ban-types
  [key in keyof T as T[key] extends Function ? never : key]: T[key];
};

export type Selectable<T> = Omit<OmitQueries<OmitMethods<T>>, '_id' | '_version'>;

export type Select<T, K extends keyof T> = Pick<T, K> & Queries<T> & XataRecord;

export type Include<T> = {
  [key in keyof T as T[key] extends XataRecord ? key : never]?: boolean | Array<keyof Selectable<T[key]>>;
};

type SortDirection = 'asc' | 'desc';

type Operator =
  | '_gt'
  | '_lt'
  | '_gte'
  | '_lte'
  | '_exists'
  | '_notExists'
  | '_endsWith'
  | '_startsWith'
  | '_pattern'
  | '_isNot'
  | '_includes'
  | '_includesSubstring'
  | '_includesPattern'
  | '_includesAll';

// TODO: restrict constraints depending on type?
// E.g. startsWith cannot be used with numbers
type Constraint<T> = Partial<Record<Operator, T>>;

type ComparableType = number | Date;

export const gt = <T extends ComparableType>(value: T): Constraint<T> => ({ _gt: value });
export const gte = <T extends ComparableType>(value: T): Constraint<T> => ({ _gte: value });
export const lt = <T extends ComparableType>(value: T): Constraint<T> => ({ _lt: value });
export const lte = <T extends ComparableType>(value: T): Constraint<T> => ({ _lte: value });
export const exists = (column: string): Constraint<string> => ({ _exists: column });
export const notExists = (column: string): Constraint<string> => ({ _notExists: column });
export const startsWith = (value: string): Constraint<string> => ({ _startsWith: value });
export const endsWith = (value: string): Constraint<string> => ({ _endsWith: value });
export const pattern = (value: string): Constraint<string> => ({ _pattern: value });
export const isNot = <T>(value: T): Constraint<T> => ({ _isNot: value });

// TODO: these can only be applied to columns of type "multiple"
export const includes = (value: string): Constraint<string> => ({ _includes: value });
export const includesSubstring = (value: string): Constraint<string> => ({ _includesSubstring: value });
export const includesPattern = (value: string): Constraint<string> => ({ _includesPattern: value });
export const includesAll = (value: string): Constraint<string> => ({ _includesAll: value });

type FilterConstraints<T> = {
  [key in keyof T]?: T[key] extends Record<string, any> ? FilterConstraints<T[key]> : T[key] | Constraint<T[key]>;
};

type BulkQueryOptions<T> = {
  filter?: FilterConstraints<T>;
  sort?:
    | {
        column: keyof T;
        direction?: SortDirection;
      }
    | keyof T;
};

type QueryOrConstraint<T, R> = Query<T, R> | Constraint<T>;

export class Query<T, R = T> {
  table: string;
  repository: Repository<T>;

  readonly _any?: QueryOrConstraint<T, R>[];
  readonly _all?: QueryOrConstraint<T, R>[];
  readonly _not?: QueryOrConstraint<T, R>[];
  readonly _none?: QueryOrConstraint<T, R>[];
  readonly _sort?: Record<string, SortDirection>;

  constructor(repository: Repository<T> | null, table: string, data: Partial<Query<T, R>>, parent?: Query<T, R>) {
    if (repository) {
      this.repository = repository;
    } else {
      this.repository = this as any;
    }
    this.table = table;

    // For some reason Object.assign(this, parent) didn't work in this case
    // so doing all this manually:
    this._any = parent?._any;
    this._all = parent?._all;
    this._not = parent?._not;
    this._none = parent?._none;
    this._sort = parent?._sort;

    Object.assign(this, data);
    // These bindings are used to support deconstructing
    // const { any, not, filter, sort } = xata.users.query()
    this.any = this.any.bind(this);
    this.all = this.all.bind(this);
    this.not = this.not.bind(this);
    this.filter = this.filter.bind(this);
    this.sort = this.sort.bind(this);
    this.none = this.none.bind(this);
  }

  any(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        _any: (this._any || []).concat(queries)
      },
      this
    );
  }

  all(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        _all: (this._all || []).concat(queries)
      },
      this
    );
  }

  not(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        _not: (this._not || []).concat(queries)
      },
      this
    );
  }

  none(...queries: Query<T, R>[]): Query<T, R> {
    return new Query<T, R>(
      this.repository,
      this.table,
      {
        _none: (this._none || []).concat(queries)
      },
      this
    );
  }

  filter(constraints: FilterConstraints<T>): Query<T, R>;
  filter<F extends keyof T>(column: F, value: FilterConstraints<T[F]> | Constraint<T[F]>): Query<T, R>;
  filter(a: any, b?: any): Query<T, R> {
    if (arguments.length === 1) {
      const constraints = a as FilterConstraints<T>;
      const queries: QueryOrConstraint<T, R>[] = [];
      for (const [column, constraint] of Object.entries(constraints)) {
        queries.push({ [column]: constraint });
      }
      return new Query<T, R>(
        this.repository,
        this.table,
        {
          _any: (this._any || []).concat(queries)
        },
        this
      );
    } else {
      const column = a as keyof T;
      const value = b as Partial<T[keyof T]> | Constraint<T[keyof T]>;
      return new Query<T, R>(
        this.repository,
        this.table,
        {
          _any: (this._any || []).concat({ [column]: value })
        },
        this
      );
    }
  }

  sort<F extends keyof T>(column: F, direction: SortDirection): Query<T, R> {
    const sort = { ...this._sort, [column]: direction };
    const q = new Query<T, R>(
      this.repository,
      this.table,
      {
        _sort: sort
      },
      this
    );

    return q;
  }

  // TODO: pagination. Maybe implement different methods for different type of paginations
  // and one to simply get the first records returned by the query with no pagination.
  async getMany(options?: BulkQueryOptions<T>): Promise<R[]> {
    // TODO: use options
    return this.repository.query(this);
  }

  async getOne(options?: BulkQueryOptions<T>): Promise<R | null> {
    // TODO: use options
    const arr = await this.getMany(); // TODO, limit to 1
    return arr[0] || null;
  }

  async deleteAll(): Promise<number> {
    // Return number of affected rows
    return 0;
  }

  include(columns: Include<T>) {
    // TODO
    return this;
  }

  toJSON() {
    const _filter = {
      _any: this._any,
      _all: this._all,
      _not: this._not,
      _none: this._none
    };
    return {
      _filter: Object.values(_filter).some(Boolean) ? _filter : undefined,
      _sort: this._sort
    };
  }
}

export abstract class Repository<T> extends Query<T, Selectable<T>> {
  select<K extends keyof Selectable<T>>(...columns: K[]) {
    return new Query<T, Select<T, K>>(this.repository, this.table, {});
  }

  abstract create(object: Selectable<T>): Promise<T>;

  abstract read(id: string): Promise<T | null>;

  abstract update(id: string, object: Partial<T>): Promise<T>;

  abstract delete(id: string): void;

  // Used by the Query object internally
  abstract query<R>(query: Query<T, R>): Promise<R[]>;
}

export class RestRepository<T> extends Repository<T> {
  client: BaseClient<any>;
  fetch: any;

  constructor(client: BaseClient<any>, table: string) {
    super(null, table, {});
    this.client = client;

    const { fetch } = client.options;

    if (fetch) {
      this.fetch = fetch;
    } else if (typeof window === 'object') {
      this.fetch = window.fetch;
    } else if (typeof require === 'function') {
      try {
        this.fetch = require('node-fetch');
      } catch (err) {
        try {
          this.fetch = require('cross-fetch');
        } catch (err) {
          throw new Error('No fetch implementation found. Please provide one in the constructor');
        }
      }
    }

    Object.defineProperty(this, 'client', { enumerable: false });
    Object.defineProperty(this, 'fetch', { enumerable: false });
    Object.defineProperty(this, 'hostname', { enumerable: false });
  }

  async request(method: string, path: string, body?: unknown) {
    const { url: xatabaseURL } = this.client.options;
    const resp: Response = await this.fetch(`${xatabaseURL}${path}`, {
      method,
      headers: {
        Accept: '*/*',
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.client.options.token}`
      },
      body: JSON.stringify(body)
    });
    if (!resp.ok) {
      try {
        const json = await resp.json();
        const message = json.message;
        if (typeof message === 'string') {
          throw new XataError(message, resp.status);
        }
      } catch (err) {
        if (err instanceof XataError) throw err;
        // Ignore errors for other reasons.
        // For example if the response's body cannot be parsed as JSON
      }
      throw new XataError(resp.statusText, resp.status);
    }
    if (resp.status === 204) return;
    return resp.json();
  }

  select<K extends keyof T>(...columns: K[]) {
    return new Query<T, Select<T, K>>(this.repository, this.table, {});
  }

  async create(object: T): Promise<T> {
    const obj = await this.request('POST', `/tables/${this.table}/data`, object);
    return this.client.initObject(this.table, obj);
  }

  async read(id: string): Promise<T | null> {
    try {
      const obj = await this.request('GET', `/tables/${this.table}/data/${id}`);
      return this.client.initObject(this.table, obj);
    } catch (err) {
      if ((err as XataError).status === 404) return null;
      throw err;
    }
  }

  async update(id: string, object: Partial<T>): Promise<T> {
    const obj = await this.request('PUT', `/tables/${this.table}/data/${id}`, object);
    return this.client.initObject(this.table, obj);
  }

  async delete(id: string) {
    await this.request('DELETE', `/tables/${this.table}/data/${id}`);
  }

  async query<R>(query: Query<T, R>): Promise<R[]> {
    const result = await this.request('POST', `/tables/${this.table}/query`, query);
    return result.records.map((record: object) => this.client.initObject(this.table, record));
  }
}

interface RepositoryFactory {
  createRepository<T>(client: BaseClient<any>, table: string): Repository<T>;
}

export class RestRespositoryFactory implements RepositoryFactory {
  createRepository<T>(client: BaseClient<any>, table: string): Repository<T> {
    return new RestRepository<T>(client, table);
  }
}

export type XataClientOptions = {
  fetch?: unknown;
  url: string;
  token: string;
  repositoryFactory?: RepositoryFactory;
};

export class BaseClient<D extends Record<string, Repository<any>>> {
  options: XataClientOptions;
  private links: Links;
  db!: D;

  constructor(options: XataClientOptions, links: Links) {
    this.options = options;
    this.links = links;
  }

  public initObject<T>(table: string, object: object) {
    const o: Record<string, unknown> = {};
    Object.assign(o, object);

    const tableLinks = this.links[table] || [];
    for (const link of tableLinks) {
      const [field, linkTable] = link;
      const value = o[field];

      if (value && typeof value === 'object') {
        const { _id } = value as any;
        if (Object.keys(value).find((col) => !col.startsWith('_'))) {
          o[field] = this.initObject(linkTable, value);
        } else if (_id) {
          o[field] = {
            _id,
            get: () => {
              this.db[linkTable].read(_id);
            }
          };
        }
      }
    }

    const db = this.db;
    o.read = function () {
      return db[table].read(o['_id'] as string);
    };
    o.update = function (data: any) {
      return db[table].update(o['_id'] as string, data);
    };
    o.delete = function () {
      return db[table].delete(o['_id'] as string);
    };

    for (const prop of ['read', 'update', 'delete']) {
      Object.defineProperty(o, prop, { enumerable: false });
    }

    // TODO: links and rev links

    Object.freeze(o);
    return o as T;
  }
}

export class XataError extends Error {
  readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }
}

export type Links = Record<string, Array<string[]>>;
