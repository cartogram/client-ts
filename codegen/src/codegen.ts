import * as fs from 'fs/promises';
import * as path from 'path';
import { singular } from 'pluralize';
import { ZodError } from 'zod';
import { Column, fileSchema, Table } from './schema';
import { join } from 'path';

import prettier from 'prettier';
import { getLanguageFromExtension } from './getLanguageFromExtension';
import { isExtensionValid } from './isExtensionValid';
import { errors } from './errors';

type GenerateOptions = {
  xataDirectory: string;
  outputFilePath: string;
  writeFile?: typeof fs.writeFile;
};

function getTypeName(tableName: string) {
  const snglr = singular(tableName);
  return snglr.substring(0, 1).toUpperCase() + snglr.substring(1);
}

function generateTableType(table: Table) {
  const { columns } = table;
  const revLinks: { table: string }[] = []; // table.rev_links || [];
  return `export interface ${getTypeName(table.name)} extends XataRecord {
    ${columns.map((column) => generateColumnType(column)).join('\n')}
    ${revLinks.map((link) => `${link.table}: Query<${getTypeName(link.table)}>`).join('\n')}
  };
  `;
}

function generateColumnType(column: Column) {
  return `${column.name}?: ${getTypeScriptType(column)}`;
}

function generateJSdocType(table: Table) {
  const { columns } = table;
  const revLinks: { table: string }[] = []; // table.rev_links || [];
  const typeName = getTypeName(table.name);
  return `
/**
 * @typedef {Object} ${typeName}
 * @property {string} id
 * @property {Object} xata
 * @property {() => Promise<${typeName}>} read
 * @property {() => Promise<${typeName}>} update
 * @property {() => Promise<void>} delete
 ${columns.map((column) => generateJSDocColumnType(column)).join('\n ')}
 ${revLinks.map((link) => `${link.table}: Query<${getTypeName(link.table)}>`).join('\n ')}
 */`;
}

function generateJSDocColumnType(column: Column) {
  return `* @property {${getTypeScriptType(column)}=} ${column.name}`;
}

function getTypeScriptType(column: Column): string {
  if (column.type === 'email') return 'string';
  if (column.type === 'text') return 'string';
  if (column.type === 'string') return 'string';
  if (column.type === 'multiple') return 'string[]';
  if (column.type === 'bool') return 'boolean';
  if (column.type === 'int') return 'number';
  if (column.type === 'link') {
    if (!column.link?.table) return 'object';
    return getTypeName(column.link.table);
  }
  if (column.type === 'object') {
    const columns = column.columns || [];
    return `{ ${columns.map((column) => generateColumnType(column)).join('; ')} }`;
  }
  return 'unknown';
}

async function readSchema(fullPath: string) {
  try {
    return await fs.readFile(fullPath, 'utf-8');
  } catch (err) {
    console.error('Could not read schema file at', fullPath);
    process.exit(1);
  }
}

function parseSchema(input: string) {
  try {
    return fileSchema.parse(JSON.parse(input));
  } catch (err) {
    console.error('The content of the schema file is not valid:');
    if (err instanceof ZodError) {
      const zodError = err as ZodError;
      for (const error of zodError.errors) {
        console.error(`  [${error.code}]`, error.message, 'at', `"${error.path.join('.')}"`);
      }
    } else {
      console.error(' ', err instanceof Error ? err.message : err);
    }
    process.exit(1);
  }
}

export type Language = 'typescript' | 'javascript';

export async function generate({ outputFilePath: output, xataDirectory, writeFile = fs.writeFile }: GenerateOptions) {
  const fullOutputPath = path.resolve(process.cwd(), output);
  const [extension] = fullOutputPath.split('.').slice(-1);

  if (!isExtensionValid(extension)) {
    throw new Error(errors.invalidCodegenOutputExtension);
  }

  const schemaFile = join(xataDirectory, 'schema.json');
  const input = await readSchema(schemaFile);
  const schema = parseSchema(input);

  const { tables } = schema;
  const links: Record<string, string[][]> = {};
  for (const table of tables) {
    links[table.name] = [];
    for (const column of table.columns) {
      if (column.link) {
        links[table.name].push([column.name, column.link.table]);
      }
    }
  }

  if (getLanguageFromExtension(extension) === 'typescript') {
    const code = `
    import {
      BaseClient,
      Query,
      Repository,
      RestRespositoryFactory,
      XataClientOptions,
      XataRecord
    } from '../../client/src';

    ${tables.map((table) => generateTableType(table)).join('\n')}

    const links = ${JSON.stringify(links)};

    export class XataClient extends BaseClient<{
      ${tables.map((table) => `"${table.name}": Repository<${getTypeName(table.name)}>;`).join('\n')}
    }> {
      constructor(options: XataClientOptions) {
        super(options, links);
        const factory = options.repositoryFactory || new RestRespositoryFactory();
        this.db = {
          ${tables.map((table) => `"${table.name}": factory.createRepository(this, "${table.name}"),`).join('\n')}
        };
      }
    }
  `;

    const pretty = prettier.format(code, { parser: 'typescript' });
    await writeFile(fullOutputPath, pretty);
    return;
  }

  const code = `
    /** @typedef { import('../../client/src').Repository } Repository */
    import {
      BaseClient,
      Query,
      RestRespositoryFactory
    } from '../../client/src';

    ${tables.map((table) => generateJSdocType(table)).join('\n')}

    const links = ${JSON.stringify(links)};

    export class XataClient extends BaseClient {
      constructor(options) {
        super(options, links);
        const factory = options.repositoryFactory || new RestRespositoryFactory();
        /** @type {{ ${tables.map((table) => `"${table.name}": Repository`).join('; ')} }} */
        this.db = {
          ${tables.map((table) => `"${table.name}": factory.createRepository(this, "${table.name}"),`).join('\n')}
        };
      }
    }
  `;

  const pretty = prettier.format(code, { parser: 'babel' });
  await writeFile(fullOutputPath, pretty);
}
