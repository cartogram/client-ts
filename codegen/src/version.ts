export const CODEGEN_VERSION = "0.2.2";
