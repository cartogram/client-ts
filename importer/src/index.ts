// These options should be common for all formats.
// Some formats may ignore some options though.
export type ParseOptions = {
  types?: string[];
  columns?: string[];
  batchSize?: number;
  maxRows?: number;
  noheader?: boolean;
  callback: (lines: string[][], columns?: string[]) => Promise<boolean | void>;
};
