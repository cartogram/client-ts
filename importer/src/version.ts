export const IMPORTER_VERSION = '0.0.0';
