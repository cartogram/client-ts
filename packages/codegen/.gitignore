# Auto generated version file
src/version.ts
