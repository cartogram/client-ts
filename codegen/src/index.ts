#!/usr/bin/env node
import { program } from 'commander';
import { join } from 'path';

import { generate } from './codegen';

program
  .name('xata-codegen')
  .description('The Xata SDK CLI is used to generate type-safe and predictable clients used to interact with Xata.')
  // There's nothing unsafe here, so disabling this rule in this case...
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  .version(require('../package.json').version);

program
  .command('generate')
  .description('Generate code from a given Xata schema.')
  .argument(
    '[schema file]',
    `A path to your local Xata schema. If you don't have this, run the pull command on this CLI first.`,
    join(process.cwd(), 'xata', 'schema.json')
  )
  .option('-o, --out <path>', 'A path to store your generated API client.', join(process.cwd(), 'XataClient'))
  .option(
    '-l, --lang [js|ts]',
    "An option to choose the type of code you'd like us to output: TypeScript (ts, preferred) or JavaScript (js)",
    'typescript'
  )
  .action((schema, { out, lang }) => {
    generate(schema, out, lang).catch(console.error);
  });

program.parse();
