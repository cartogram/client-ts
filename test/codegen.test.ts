import { afterEach, describe, expect, it, vi } from 'vitest';
import { generate } from '../packages/codegen/src/codegen';
import { VERSION as CODEGEN_VERSION } from '../packages/codegen/src/version';

afterEach(() => {
  vi.clearAllMocks();
});

describe('generate', () => {
  it('should respect numbers in names', async () => {
    const output = await generate({
      schema: {
        formatVersion: '1.0',
        tables: [
          {
            name: '1teams-case',
            columns: [
              { name: '2nameCase', type: 'string' },
              { name: '3Labels', type: 'multiple' }
            ]
          }
        ]
      },
      language: 'typescript',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname'
    });

    expect(stableVersion(output.transpiled)).toMatchSnapshot();
  });

  it('should respect case naming', async () => {
    const output = await generate({
      schema: {
        formatVersion: '1.0',
        tables: [
          {
            name: 'teams_Like',
            columns: [
              { name: 'name-test', type: 'string' },
              { name: 'labels_Test', type: 'multiple' },
              { name: 'ownerFoo', type: 'link', link: { table: 'users-foo' } }
            ]
          },
          {
            name: 'users-foo',
            columns: [
              { name: 'email-random', type: 'email' },
              { name: 'full_name', type: 'string' },
              { name: 'teamLike', type: 'link', link: { table: 'teams_Like' } }
            ]
          }
        ]
      },
      language: 'typescript',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname'
    });

    expect(stableVersion(output.transpiled)).toMatchSnapshot();
  });

  it('should inject branch if passed', async () => {
    const output = await generate({
      schema: {
        formatVersion: '1.0',
        tables: [
          {
            name: 'users',
            columns: [{ name: 'name', type: 'string' }]
          }
        ]
      },
      language: 'typescript',
      databaseURL: 'https://workspace-1234.xata.sh/db/dbname',
      branch: 'feature-branch'
    });

    expect(stableVersion(output.transpiled)).toMatchSnapshot();
  });
});

// This method will replace the current version of the codegen with a placeholder so snapshots don't change
// if we just bump the codegen version
function stableVersion(code: string) {
  return code.replace(
    `Generated by Xata Codegen ${CODEGEN_VERSION}. Please do not edit`,
    'Generated by Xata Codegen CODEGEN_VERSION. Please do not edit'
  );
}
