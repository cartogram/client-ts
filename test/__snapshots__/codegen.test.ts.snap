// Vitest Snapshot v1

exports[`generate > should generate CJS code 1`] = `
"// Generated by Xata Codegen CODEGEN_VERSION. Please do not edit.
const {
  BaseClientOptions,
  buildClient,
  SchemaInference,
  XataRecord,
} = require(\\"@xata.io/client\\");
/** @typedef { import('./types').SchemaTables } SchemaTables */
/** @type { SchemaTables } */
const tables = [{ name: \\"users\\", columns: [{ name: \\"name\\", type: \\"string\\" }] }];
/** @type { import('@xata.io/client').ClientConstructor<{}> } */
const DatabaseClient = buildClient();
const defaultOptions = {
  databaseURL: \\"https://workspace-1234.xata.sh/db/dbname\\",
  branch: \\"feature-branch\\",
};
/** @typedef { import('./types').DatabaseSchema } DatabaseSchema */
/** @extends DatabaseClient<DatabaseSchema> */
export class XataClient extends DatabaseClient {
  constructor(options) {
    super({ ...defaultOptions, ...options }, tables);
  }
}
let instance = undefined;
/** @type { () => XataClient } */
export const getXataClient = () => {
  if (instance) return instance;
  instance = new XataClient();
  return instance;
};
"
`;

exports[`generate > should ignore moduleType for TS code 1`] = `
"// Generated by Xata Codegen CODEGEN_VERSION. Please do not edit.
import {
  BaseClientOptions,
  buildClient,
  SchemaInference,
  XataRecord,
} from \\"@xata.io/client\\";

const tables = [
  { name: \\"users\\", columns: [{ name: \\"name\\", type: \\"string\\" }] },
] as const;

export type SchemaTables = typeof tables;
export type InferredTypes = SchemaInference<SchemaTables>;

export type Users = InferredTypes[\\"users\\"];
export type UsersRecord = Users & XataRecord;

export type DatabaseSchema = {
  users: UsersRecord;
};

const DatabaseClient = buildClient();

const defaultOptions = {
  databaseURL: \\"https://workspace-1234.xata.sh/db/dbname\\",
  branch: \\"feature-branch\\",
};

export class XataClient extends DatabaseClient<DatabaseSchema> {
  constructor(options?: BaseClientOptions) {
    super({ ...defaultOptions, ...options }, tables);
  }
}

let instance: XataClient | undefined = undefined;

/** @type { () => XataClient } */
export const getXataClient = () => {
  if (instance) return instance;

  instance = new XataClient();
  return instance;
};
"
`;

exports[`generate > should inject branch if passed 1`] = `
"// Generated by Xata Codegen CODEGEN_VERSION. Please do not edit.
import {
  BaseClientOptions,
  buildClient,
  SchemaInference,
  XataRecord,
} from \\"@xata.io/client\\";

const tables = [
  { name: \\"users\\", columns: [{ name: \\"name\\", type: \\"string\\" }] },
] as const;

export type SchemaTables = typeof tables;
export type InferredTypes = SchemaInference<SchemaTables>;

export type Users = InferredTypes[\\"users\\"];
export type UsersRecord = Users & XataRecord;

export type DatabaseSchema = {
  users: UsersRecord;
};

const DatabaseClient = buildClient();

const defaultOptions = {
  databaseURL: \\"https://workspace-1234.xata.sh/db/dbname\\",
  branch: \\"feature-branch\\",
};

export class XataClient extends DatabaseClient<DatabaseSchema> {
  constructor(options?: BaseClientOptions) {
    super({ ...defaultOptions, ...options }, tables);
  }
}

let instance: XataClient | undefined = undefined;

/** @type { () => XataClient } */
export const getXataClient = () => {
  if (instance) return instance;

  instance = new XataClient();
  return instance;
};
"
`;

exports[`generate > should respect case naming 1`] = `
"// Generated by Xata Codegen CODEGEN_VERSION. Please do not edit.
import {
  BaseClientOptions,
  buildClient,
  SchemaInference,
  XataRecord,
} from \\"@xata.io/client\\";

const tables = [
  {
    name: \\"teams_Like\\",
    columns: [
      { name: \\"name-test\\", type: \\"string\\" },
      { name: \\"labels_Test\\", type: \\"multiple\\" },
      { name: \\"ownerFoo\\", type: \\"link\\", link: { table: \\"users-foo\\" } },
    ],
  },
  {
    name: \\"users-foo\\",
    columns: [
      { name: \\"email-random\\", type: \\"email\\" },
      { name: \\"full_name\\", type: \\"string\\" },
      { name: \\"teamLike\\", type: \\"link\\", link: { table: \\"teams_Like\\" } },
    ],
  },
] as const;

export type SchemaTables = typeof tables;
export type InferredTypes = SchemaInference<SchemaTables>;

export type TeamsLike = InferredTypes[\\"teams_Like\\"];
export type TeamsLikeRecord = TeamsLike & XataRecord;

export type UsersFoo = InferredTypes[\\"users-foo\\"];
export type UsersFooRecord = UsersFoo & XataRecord;

export type DatabaseSchema = {
  teams_Like: TeamsLikeRecord;
  \\"users-foo\\": UsersFooRecord;
};

const DatabaseClient = buildClient();

const defaultOptions = {
  databaseURL: \\"https://workspace-1234.xata.sh/db/dbname\\",
};

export class XataClient extends DatabaseClient<DatabaseSchema> {
  constructor(options?: BaseClientOptions) {
    super({ ...defaultOptions, ...options }, tables);
  }
}

let instance: XataClient | undefined = undefined;

/** @type { () => XataClient } */
export const getXataClient = () => {
  if (instance) return instance;

  instance = new XataClient();
  return instance;
};
"
`;

exports[`generate > should respect numbers in names 1`] = `
"// Generated by Xata Codegen CODEGEN_VERSION. Please do not edit.
import {
  BaseClientOptions,
  buildClient,
  SchemaInference,
  XataRecord,
} from \\"@xata.io/client\\";

const tables = [
  {
    name: \\"1teams-case\\",
    columns: [
      { name: \\"2nameCase\\", type: \\"string\\" },
      { name: \\"3Labels\\", type: \\"multiple\\" },
    ],
  },
] as const;

export type SchemaTables = typeof tables;
export type InferredTypes = SchemaInference<SchemaTables>;

export type $1teamsCase = InferredTypes[\\"1teams-case\\"];
export type $1teamsCaseRecord = $1teamsCase & XataRecord;

export type DatabaseSchema = {
  \\"1teams-case\\": $1teamsCaseRecord;
};

const DatabaseClient = buildClient();

const defaultOptions = {
  databaseURL: \\"https://workspace-1234.xata.sh/db/dbname\\",
};

export class XataClient extends DatabaseClient<DatabaseSchema> {
  constructor(options?: BaseClientOptions) {
    super({ ...defaultOptions, ...options }, tables);
  }
}

let instance: XataClient | undefined = undefined;

/** @type { () => XataClient } */
export const getXataClient = () => {
  if (instance) return instance;

  instance = new XataClient();
  return instance;
};
"
`;
